stan/math/prim/functor/reduce_sum.hpp

```diff
@@ -20,76 +20,113 @@ namespace math {
+                            msgs, std::forward<OpArgs>(args)...);
   }
 };
 
+/**
+ * Arithmetic definition of implimentation called by reduce `reduce_sum`.
+ * @tparam ReduceFunction An type with a valid `operator()`
+ * @tparam ReturnType The return type of the call to
+ * `ReduceFunction::operator()`
+ * @tparam Vec type of the first input argument with an `operator[]`
+ * @tparam Args Parameter pack holding the types of the object send to
+ * `ReduceFunction::operator()`
+ */
 template <typename ReduceFunction, typename ReturnType, typename Vec,
           typename... Args>
 struct reduce_sum_impl<ReduceFunction, require_arithmetic_t<ReturnType>,
                        ReturnType, Vec, Args...> {
+  /**
+   * Internal object used in `tbb::parallel_reduce`
+   * @note see link [here](https://tinyurl.com/vp7xw2t) for requirements.
+   */
   struct recursive_reducer {
-    using vmapped_t = Vec;
-    const vmapped_t& vmapped_;
+    Vec vmapped_;
     std::ostream* msgs_;
-    std::tuple<const Args&...> args_tuple_;
-    return_type_t<Vec, Args...> sum_;
+    std::tuple<Args...> args_tuple_;
+    return_type_t<Vec, Args...> sum_{0.0};
 
-    recursive_reducer(const vmapped_t& vmapped, std::ostream* msgs,
-                      const Args&... args)
-        : vmapped_(vmapped), msgs_(msgs), args_tuple_(args...), sum_(0.0) {}
+    template <typename OpVec, typename... OpArgs,
+              typename = require_same_t<Vec, OpVec>,
+              typename = require_all_t<std::is_same<Args, OpArgs>...>>
+    recursive_reducer(OpVec&& vmapped, std::ostream* msgs, OpArgs&&... args)
+        : vmapped_(std::forward<OpVec>(vmapped)),
+          msgs_(msgs),
+          args_tuple_(std::forward<OpArgs>(args)...) {}
 
     recursive_reducer(recursive_reducer& other, tbb::split)
         : vmapped_(other.vmapped_),
           msgs_(other.msgs_),
-          args_tuple_(other.args_tuple_),
-          sum_(0.0) {}
+          args_tuple_(other.args_tuple_) {}
 
     void operator()(const tbb::blocked_range<size_t>& r) {
       if (r.empty()) {
         return;
       }
 
-      auto start = vmapped_.begin();
+      auto start = this->vmapped_.begin();
       std::advance(start, r.begin());
-      auto end = vmapped_.begin();
+      auto end = this->vmapped_.begin();
       std::advance(end, r.end());
 
-      const vmapped_t sub_slice(start, end);
+      const std::decay_t<Vec> sub_slice(start, end);
 
       sum_ += apply(
           [&](auto&&... args) {
             return ReduceFunction()(r.begin(), r.end() - 1, sub_slice, msgs_,
                                     args...);
           },
-          args_tuple_);
+          this->args_tuple_);
     }
-
-    void join(const recursive_reducer& child) { sum_ += child.sum_; }
+    /**
+     * Join the sum of the reducers
+     * @param child A sub-reducer that is aggregated to make the final reduced
+     * value.
+     * @note side effect of updating `recursive_reducer->sum_`
```

Comment:
This isn't a side effect, it's the effect!

Reply:
I can change this, but I've always thought of side effects as "things manipulated outside of the function that is not the return". like R's `plot()`. does that make sense?